2018/ideas-list-shogun.md

```diff
@@ -0,0 +1,552 @@
+ * Desirable: Experience with the [benchmarking system](https://github.com/zoq/benchmarks)
+
+### Details
+Here are some examples of what topics should be covered.
+
+### Runtime
+
+Have a look at benchmark comparisons of Shogun with other libraries at [mlpack's benchmarking framework](http://www.mlpack.org/benchmark.html).
+You will notices that sometimes Shogun does quite well, like for KMeans
+
+| dataset | mlpy | scikit | shogun | weka | mlpack |
+|---------|-------|-------|--------|------|--------|
+| corel-histogram | 3.59s | 0.73s | 1.11s | 19.43s | 1.92s |
+| mnist | 119.83s | 46.13s | 16.02s | 1558.07s | 61.35s |
+
+On the other hand, there are situations that are less than optimal, like for linear regression, where Shogun fails.
+
+| dataset | mlpy | scikit | shogun | weka | mlpack |
+|---------|-------|-------|--------|------|--------|
+| arcene | failure | 0.24s | failure | 3.16s | 0.42s |
+| cosExp | 0.13s | 0.08s | failure | 17.42s | 0.13s|
+
+Anotoher one is linear ridge regression, where Shogun is extremely slow
+
+| dataset |  scikit | shogun |
+|---------|---------|--------|
+| webpage | 1.94s | >9000s |
+
+Again, we don't want Shogun to be the fastest candidate everywhere. We only don't want it to be the slowest by far.
+
+### Awkward API
+Example: have a look at [GMM](shogun.ml/CGMM).
+It has 3 train methods, awkward methods like `get_nth_mean`, multiple methods to `apply` it (`::cluster, ::get_likelihood_example`), etc.
+A first step would be to rename the methods to something that looks nice, or to remove them (we have tags so no need for getters/setters anymore).
+Next, GMM is nothing else but a supervised learning algorithm, so it should support that interface: `fit, predict`, and not offer its own methods.
+Next, GMM is also a distribution that can be sampled from, so it should be possible to turn it into an API that supports sampling.
+
+We actually wrote some API desiderata for the [user experience project](GSoC_2018_project_usability), which overlaps with the project in terms of API.
+Think: you identify bad API, and how it should be instead, user experience project person implements basics for your changes to be possible, you change the algorithm.
+
+### Documentation issues
+Some bad examples:
+
+* [Example of one sentence docs](http://shogun.ml/CCCSOSVM)
+* [Example of no documentation at all](http://shogun.ml/CLibSVM)
+* [Example of bad documentation](http://shogun.ml/CLibSVMCMulticlassLibSVM) -- no description of what happens or how expensive it is.
+* [Example of bad documentation](http://shogun.ml/CRandomKitchenSinksDotFeatures) -- talks about using the features in ERM, but this class is just about feature embeddings, so it should talk about the embedding, it computational costs, and what one can do with it: pass it to linear algorithms (like a linear SVM).
+
+You get the point...
+
+## First steps
```

Comment:
Is this a new project? It has a level 2 heading.

It would be good to normalize the headings so that a level 1 or 2 heading signals a new project and the same heading names should always have the same heading level.